FEAT: added medsafetybench dataset

nthsneha opened the pull request:
# medsafetybench Dataset Integration for PyRIT 

This PR adds support for the medsafetybench dataset to PyRIT. The medsafetybench dataset is a benchmark designed to evaluate and improve the medical safety of large language models (LLMs). It includes medical safety demonstrations consisting of harmful medical prompts and corresponding safe responses. By integrating this dataset, PyRIT can be used to evaluate the safety of LLMs in high-risk healthcare contexts, helping surface potentially harmful model completions and benchmark alignment efforts.

## Key Changes 

- Added a new function fetch_medsafetybench() to pyrit.datasets.
- The function loads data from both train and test folders, including nested sources (gpt4, llama2).
- Each prompt-response pair is tagged with a default harm_categories = ["medical misinformation"].
- Added the function to doc/api.rst for inclusion in the public API reference.

## Related Issue

Closes: #948

This PR addresses the dataset integration request made by the maintainer: @romanlutz
Tagging for review.

## Tests and Documentation

### Tests: 
- Tested locally using a python script to load all examples from the dataset. Confirmed successful parsing of 1,800 examples.  
### Documentation: 
- doc/api.rst updated to include fetch_medsafetybench() in the API reference. Inline docstrings provided to explain function purpose and expected format.

## References 
https://github.com/AI4LIFE-GROUP/med-safety-bench

### Citation
@article{han2024medsafetybench,
  title={MedSafetyBench: Evaluating and Improving the Medical Safety of Large Language Models},
  author={Han, Tessa and Kumar, Aounon and Agarwal, Chirag and Lakkaraju, Himabindu},
  journal={NeurIPS},
  year={2024}
}

nthsneha commented:
@microsoft-github-policy-service agree

romanlutz commented on pyrit/datasets/fetch_medsafetybench.py:
```diff
@@ -0,0 +1,43 @@
+
+def fetch_medsafetybench(base_dir: str = "datasets") -> List[Dict]:
+    """
+    Loads the MedSafetyBench dataset from nested train and test subfolders 
+    and returns a list of dictionaries with prompt, completion, and harm_categories.
```
Can we add 
- paper title
- link to paper
- link to dataset
- list of authors
?

romanlutz commented on pyrit/datasets/fetch_medsafetybench.py:
```diff
@@ -0,0 +1,43 @@
+        }
+    """
+    prompts = []
+    for split in ["train", "test"]:
+        split_dir = os.path.join(base_dir, split)
+        if not os.path.exists(split_dir):
+            continue
+
+        for root, _, files in os.walk(split_dir): 
+            for file in files:
+                if file.endswith(".csv"):
+                    file_path = os.path.join(root, file)
+                    with open(file_path, "r", encoding="utf-8") as f:
+                        reader = csv.DictReader(f)
+                        for row in reader:
```
This assumes that we've downloaded the files. The purpose of this function is to download the dataset and provide it as SeedPromptDataset.

Based on the docs https://github.com/AI4LIFE-GROUP/med-safety-bench/tree/main?tab=readme-ov-file#dataset-structure I would say that we should allow people to pass a parameter to specify whether they want
- train
- test
- generated
- all (all 3 combined)

We can call the param `subset_name` like we did in other datasets.

romanlutz replied:
I would just go with test, train, generated, and combine the files for each category.

romanlutz commented on pyrit/datasets/fetch_medsafetybench.py:
```diff
@@ -0,0 +1,43 @@
+                            if prompt and completion:
+                                prompts.append({
+                                    "prompt": prompt,
+                                    "completion": completion,
```
The completion doesn't matter to us. We only need the prompt.

romanlutz commented on pyrit/datasets/fetch_medsafetybench.py:
```diff
@@ -0,0 +1,43 @@
+import csv
+from typing import List, Dict
+
+def fetch_medsafetybench(base_dir: str = "datasets") -> List[Dict]:
```
This should return a SeedPromptDataset, which is essentially an object containing a list of SeedPrompts. See almost any other dataset file for reference.

romanlutz commented on pyrit/datasets/fetch_medsafetybench.py:
nit: just the dataset name is sufficient for the file name. `fetch` can be omitted

romanlutz commented on pyrit/datasets/fetch_medsafetybench.py:
The __init__.py file should include the fetch function for shorter imports

romanlutz commented on pyrit/datasets/fetch_medsafetybench.py:
```diff
@@ -0,0 +1,43 @@
+                                prompts.append({
+                                    "prompt": prompt,
+                                    "completion": completion,
+                                    "harm_categories": ["medical misinformation"]
```
This isn't bad, but probably not exactly correct in all cases. Some are more in the medical safety direction. Perhaps just "medicine" or "medical safety" would be preferable?

nthsneha commented:
@romanlutz made all the required changes.

romanlutz commented:
@nthsneha it looks like there are still lots of open comments that haven't been addressed yet.

nthsneha commented:
@romanlutz 
So far, 
- I changed doc/api.rst: inserted fetch_medsafetybench in alphabetical order

- added link to paper, dataset and names of authors

- generated links for the different dataset categories: test, train, generated, all(test+train)

- changed harm_categories from "medical misinformation" to "medical safety" 

- changed the dataset name to medsafetybench.py

- imported and returned seedprompt object 

- removed completion

Is there something I should change or am i missing out on something?

romanlutz commented:
@nthsneha if you look through either this page or the "files changed" view you should see the open comment threads. It also looks like you copied the file. There are two now. Perhaps one of them has the comments addressed already?

nthsneha commented:
@romanlutz The last commit that I made addresses all the comments/issues.

romanlutz commented:
> @romanlutz The last commit that I made addresses all the comments/issues.

This is what I see in the "files changed" view:
<img width="231" height="163" alt="image" src="https://github.com/user-attachments/assets/f737bfc4-f324-4e9d-a546-f2fb5e4ba3e6" />
One of these files doesn't belong there. I'll cross-check the comments again when there's only one left. Thanks!

nthsneha commented:
@romanlutz You're right, there were 2 files. I deleted one of them.

nthsneha commented:
@romanlutz I made some changes, please review

nthsneha commented:
@romanlutz should I make any changes now?

romanlutz commented:
There is a requirement for integration tests for each fetch method. This one doesn't have one yet. Can you just add it to the list https://github.com/Azure/PyRIT/blob/ab623bdf773c52efcf3111578b95730a0cd24039/tests/integration/datasets/test_fetch_datasets.py#L64

nthsneha commented:
@romanlutz ran a pre-commit

romanlutz commented:
Yes but it failed. It needs to pass.

nthsneha commented:
@romanlutz
